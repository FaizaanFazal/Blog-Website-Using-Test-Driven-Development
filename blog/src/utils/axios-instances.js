// import axios from 'axios';

// export const axiosInstance = axios.create({
//   baseURL: 'http://localhost:8000/',
// });
import axios from 'axios';

const placeholderApi = axios.create({
  baseURL: 'http://localhost:8000',
});

export default placeholderApi;
