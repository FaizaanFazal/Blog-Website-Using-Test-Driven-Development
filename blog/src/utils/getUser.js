export const getUser = () => {
    const userString = localStorage.getItem('user');
    if (userString) {
      const user = JSON.parse(userString);
      return user;
    } else {
      return null;
    }
  };