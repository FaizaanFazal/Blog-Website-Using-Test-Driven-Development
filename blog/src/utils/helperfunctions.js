import { fireEvent } from '@testing-library/react';
import images from './images';

export const isProperImageURL = (url) => {
  /* eslint-disable-next-line no-useless-escape */
  const regex = /^(https?|ftp):\/\/(([a-z\d]([a-z\d-]*[a-z\d])?\.)+[a-z]{2,}|localhost)(\/[-a-z\d%_.~+]*)*(\?[;&a-z\d%_.~+=-]*)?(\#[-a-z\d_]*)?$/i;
  return regex.test(url);
};
export const isProperEmail = (email) => {
  const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return emailRegex.test(email);
};
export const isProperPass = (pass) => {
  const passregex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$/;
  return passregex.test(pass);
};
export function backspace(element) {
  let actuallyTyped = element.value;

  const backspaceKey = {
    key: 'Backspace',
    code: 8,
    inputType: 'deleteContentBackward',
  };

  const sharedEventConfig = {
    key: backspaceKey.key,
    charCode: backspaceKey.code,
    keyCode: backspaceKey.code,
    which: backspaceKey.code,
    modifier: backspaceKey.modifier,
  };
  const downEvent = fireEvent.keyDown(element, sharedEventConfig);

  if (downEvent) {
    actuallyTyped = actuallyTyped.slice(0, -1);

    fireEvent.input(element, {
      target: { value: actuallyTyped },
      inputType: backspaceKey.inputType,
      bubbles: true,
      cancelable: true,
    });
  }

  fireEvent.keyUp(element, sharedEventConfig);
}

export const small = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: 'Elena Rodriguez' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    verified: true,
    slug: 'post1',
    content: "test Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.",
  },
];
export const greater = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: ' Elena' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    slug: 'post1',
    verified: true,
    content: "Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n",
  }];
