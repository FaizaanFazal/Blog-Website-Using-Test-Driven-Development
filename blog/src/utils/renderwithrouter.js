import {render} from '@testing-library/react'
import * as userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom'
import {MemoryRouter} from 'react-router-dom'

export const renderWithRouter = (ui, {route = '/'} = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    user: userEvent.setup,
    ...render(ui, {wrapper: MemoryRouter}),
  }
}