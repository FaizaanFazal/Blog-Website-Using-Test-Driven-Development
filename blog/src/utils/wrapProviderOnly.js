import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

const AllTheProviders = ({ children }) => (
  <Provider store={reduxStore}>
    {children}
  </Provider>
);

export const renderWithProvidersOnly = (ui, options) => render(<AllTheProviders>{ui}</AllTheProviders>, options);

// export function renderWithProviders(
//   ui,
//   {
//     // preloadedState = {},
//     // Automatically create a store instance if no store was passed in
//     store = reduxStore,
//     ...renderOptions
//   } = {},
// ) {
//   function Wrapper({ children }) {
//     return <Provider store={store}>{children}</Provider>;
//   }

//   // Return an object with the store and all of RTL's query functions
//   return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
// }

renderWithProvidersOnly.propTypes = {
  children: PropTypes.node.isRequired,
};
