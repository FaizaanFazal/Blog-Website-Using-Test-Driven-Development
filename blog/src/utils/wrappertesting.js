import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

export function renderWithProviders(
  ui,
  {
    // preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = reduxStore,
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

renderWithProviders.propTypes = {
  children: PropTypes.node.isRequired,
};
