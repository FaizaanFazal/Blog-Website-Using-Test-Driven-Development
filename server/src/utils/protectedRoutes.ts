import { authenticateToken } from "@/models/authModel";
import { Request, Response, NextFunction, RequestHandler } from "express";


export const protectedRoutes = [
    {
        method: "",
        path: ""
    },
    {
        method: "",
        path: ""
    }];


//check protected routes to check authentication
export const isProtectedRouteMiddleware: RequestHandler = (req: Request, res: Response, next: NextFunction) => {
    console.log(req.method)
    const protectedPathData = protectedRoutes.find((route) => route.path === req.path);
    if (protectedPathData) {
        const authenticatedUser = authenticateToken(req, res, next);
        if (authenticatedUser !== 403 && authenticatedUser !== 401) {
            req.user = authenticatedUser;
            console.log("User", authenticatedUser)
        } else {
            res.status(authenticatedUser);
            res.end();
        }
    } else {
        next();
    }
};