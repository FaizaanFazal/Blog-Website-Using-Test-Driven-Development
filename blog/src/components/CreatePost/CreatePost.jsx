// eslint-disable jsx-a11y/label-has-associated-control
import React, { useEffect, useState } from 'react';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import '../../styles/Card.scss';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { isProperImageURL } from '../../utils/helperfunctions';
import { createBlogs } from '../../Redux/blogSlice';

export default function CreatePost() {
  const user = useSelector((state) => state.user);
  const [title, setTitle] = useState('');
  const [imgUrl, setImgUrl] = useState('');
  const [imgAlt, setImgAlt] = useState('');
  const [content, setContent] = useState('');
  const [slug, setSlug] = useState('');
  const [imgdisplay, setImgdisplay] = useState({ src: '', status: false });
  const [errorTitle, setErrorTitle] = useState('');
  const [errorImgurl, setErrorImgurl] = useState('');
  const [errorImgalt, setErrorImgalt] = useState('');
  const [errorContent, setErrorContent] = useState('');
  const [errorSlug, setErrorSlug] = useState('');

  const navigate = useNavigate();
  const dispatch = useDispatch();

  const titleValidation = (e) => {
    setTitle(e.target.value);
    const { value } = e.target;
    if (value.length < 8) { setErrorTitle('* Required at least 8 characters.'); } else { setErrorTitle(null); }
  };
  // const fetchImage = async () => {
  //   try {
  //     const response = await fetch(imgUrl,{
  //       headers: {
  //         Accept: "image/*",
  //       },
  //     });
  //     console.log(response.headers.get("Content-Type"))
  //     console.log(response.status )
  //     if (response.headers.get("Content-Type").startsWith("image/jpeg")) {
  //       setImgdisplay(true)
  //     } else {
  //       setImgdisplay(false)
  //     }
  //   } catch (error) {
  //     console.log(error)
  //     setImgdisplay(false)
  //   }
  // };

  const imgurlValidation = (e) => {
    const { value } = e.target;
    const check = isProperImageURL(value);
    setImgUrl(value);
    if (value.length < 8) { setErrorImgurl('* Required at least 8 characters.'); }
    if (check === false) { setErrorImgurl('* Incorrect URL format example:(https://google.com)'); } else { setErrorImgurl(null); }
    if (check === true) { setImgdisplay({ src: value, status: true }); } else { setImgdisplay({ src: '', status: false }); }
  };

  const imgaltValidation = (e) => {
    const { value } = e.target;
    setImgAlt(value);
    if (value.length < 8) { setErrorImgalt('* Required at least 8 characters.'); }
  };

  const contentValidation = (e) => {
    setContent(e.target.value);
    const { value } = e.target;
    if (value.length < 100) { setErrorContent('* Required at least 100 characters.'); } else { setErrorContent(null); }
  };

  const slugValidation = (e) => {
    setSlug(e.target.value);
    const { value } = e.target;
    if (value.length < 5) { setErrorSlug('* Required at least 5 characters.'); } else { setErrorSlug(null); }
  };

  const validateAll = () => {
    if (!title || title.length < 8) {
      toast.error('title incorrect: min length 8', { position: toast.POSITION.TOP_RIGHT, duration: 5000 });
      return false;
    }
    const c = isProperImageURL(imgUrl);
    if (!imgUrl || imgUrl.length < 8 || c === false) {
      toast.error('Image Url incorrect', { position: toast.POSITION.TOP_RIGHT, duration: 5000 });
      return false;
    }
    if (!content || content.length < 100) {
      toast.error('content incorrect: min length 100', { position: toast.POSITION.TOP_RIGHT, duration: 5000 });
      return false;
    }
    if (!slug || slug.length < 5) {
      toast.error('slug incorrect: min length 5', { position: toast.POSITION.TOP_RIGHT, duration: 5000 });
      return false;
    }
    return true;
  };

  const handlesubmit = async (e) => {
    e.preventDefault();
    const all = validateAll();
    if (all === false) {
      return false;
    }
    let user = localStorage.getItem('user');
    user = JSON.parse(user);
    console.log(user);
    const blog = {
      authorId: user.id,
      title,
      imageSrc: imgUrl,
      imageAlt: imgAlt,
      content,
      slug,
    };
    dispatch(createBlogs(blog)).then((result) => {
      if (result.payload) {
        toast.success('Blog added succesfully', { position: toast.POSITION.TOP_RIGHT, autoClose: 2000 });
        setTitle('');
        setImgUrl('');
        setImgAlt('');
        setContent('');
        setSlug('');
      } else {
        toast.error(result.error.message, { position: toast.POSITION.TOP_RIGHT, autoClose: 2000 });
      }
    });
    return true;
  };

  const areFieldfilled = () => {
    if (!title || !imgUrl || !content || !slug) {
      return false;
    }
    return true;
  };

  useEffect(() => {
    const userr = localStorage.getItem('user');
    if (userr === null) {
      console.log(userr);
      navigate('/login');
    }
  }, [user]);

  return (
    <section data-testid="createpost">
      <ToastContainer />
      <div className="container mb-5">
        <h4 className='"title title-lg"'>Write a Post</h4>
        <form className="text-black">
          <div className="form-group">
            <label>Title</label>
            <input
              type="text"
              data-testid="inputtitle"
              className="form-control"
              placeholder="Title..."
              value={title}
              onChange={titleValidation}
            />
            {errorTitle && <small data-testid="errorTitle" style={{ color: 'red' }}>{errorTitle}</small>}
          </div>

          {imgdisplay.status === true
            && (
              <div className="customing" data-testid="ImageCont">
                <img
                  data-testid="Image"
                  src={imgdisplay.src || null}
                  alt={title}
                  className="object-fit-cover"
                  onError={({ currentTarget }) => {
                    // eslint-disable-next-line no-param-reassign
                    currentTarget.onerror = null; // prevents looping
                    // eslint-disable-next-line no-param-reassign
                    currentTarget.src = 'https://media.istockphoto.com/id/1456566772/photo/page-not-found-404-design-404-error-web-page-concept-on-a-computer-screen-3d-illustration.jpg?s=1024x1024&w=is&k=20&c=5gcQ1JbTHMqwEpw13pzVxpR8ajQ1gsrxuLOmd4CMtro=';
                  }}
                />
              </div>
            )}

          <div className="form-group">
            <label>Image Url</label>
            <input
              type="url"
              data-testid="imgInput"
              className="form-control"
              placeholder="Image Upload"
              value={imgUrl}
              onChange={imgurlValidation}
            />
            {errorImgurl && (
              <small data-testid="errorImg" style={{ color: 'red' }}>
                {errorImgurl}
                {' '}
                <br />
              </small>
            )}
          </div>

          <div className="form-group">
            <label>Image Alt</label>
            <input
              type="text"
              data-testid="imgaltinput"
              className="form-control"
              placeholder="Image Alt"
              value={imgAlt}
              onChange={imgaltValidation}
            />
            {errorImgurl && (
              <small data-testid="errorImgalt" style={{ color: 'red' }}>
                {errorImgalt}
                {' '}
                <br />
              </small>
            )}
            <small className="text-warning">*Note if image is not displayed after url insertion then check your url!! </small>
          </div>

          <div className="form-group">
            <label>Content</label>
            <textarea
              rows={3}
              data-testid="inputcontent"
              className="form-control"
              placeholder="Content...."
              value={content}
              onChange={contentValidation}
            />
            {errorContent && <small data-testid="errorContent" style={{ color: 'red' }}>{errorContent}</small>}
          </div>

          <div className="form-group">
            <label>Slug</label>
            <input
              type="text"
              data-testid="inputslug"
              className="form-control col-sm-6"
              placeholder="Slug must be unique for post"
              value={slug}
              onChange={slugValidation}
            />
            {errorSlug && <small data-testid="errorSlug" style={{ color: 'red' }}>{errorSlug}</small>}
          </div>

          <div className="text-center">
            <button
              data-testid="submitbtn"
              type="submit"
              className="btn btn-primary rounded-pill"
              onClick={handlesubmit}
              disabled={!areFieldfilled()}
            >
              Submit
            </button>
          </div>
        </form>
      </div>

    </section>
  );
}
