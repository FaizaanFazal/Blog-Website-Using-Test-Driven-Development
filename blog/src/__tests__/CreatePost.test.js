import { fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from '../utils/wrappertesting';
import CreatePost from "../components/CreatePost/CreatePost"

describe('Unit tests for CreatePost component', () => {
  it('Dom test for rendering CreatePost', () => {
    renderWithProviders(<CreatePost />);
    const featureElement = screen.getByTestId('createpost');
    expect(featureElement).toBeInTheDocument();
  });

  it('snapshot test to check CreatePost component', () => {
    const { asFragment } = renderWithProviders(<CreatePost />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Updates input value on change in title', () => {
    renderWithProviders(<CreatePost />);
    const inputTitle = screen.getByTestId('inputtitle');
    fireEvent.change(inputTitle, { target: { value: 'test' } });
    expect(inputTitle.value).toBe('test');
  });

  it('Updates input value on change in Content', () => {
    renderWithProviders(<CreatePost />);
    const inputcontent = screen.getByTestId('inputcontent');
    fireEvent.change(inputcontent, { target: { value: 'test' } });
    expect(inputcontent.value).toBe('test');
  });

  it('Submit button is rendered', () => {
    renderWithProviders(<CreatePost />);
    const Submitbtn = screen.getByTestId('submitbtn');
    expect(Submitbtn).toBeInTheDocument();
  });

});
