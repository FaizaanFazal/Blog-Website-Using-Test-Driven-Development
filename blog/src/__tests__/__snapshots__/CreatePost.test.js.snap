// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for CreatePost component snapshot test to check CreatePost component 1`] = `
<DocumentFragment>
  <div>
    CreatePost
  </div>
</DocumentFragment>
`;
